### ISSUE TITLE
CCLI Song Search & Reporting Script

### ISSUE THREAD

**Issue by vassbo** [ChurchApps/FreeShow#883]
**Describe the feature**
Log the song usage to a data file that can be requested.
More details here: https://github.com/ChurchApps/FreeShow/issues/879#issuecomment-2406928189

**Additional context**
Add a more automated way of uploading data to CCLI.

**Comment by vassbo** [comment]
Format suggestion:
```json
[{
  "name": "Lord for the years",
  "time": 1728637038559,
  "metadata": { "author": "Timothy Dudley Smith", "CCLI":  3274952 },
}, ...]
```

Do we need more info in the log @yurividal?

**Comment by yurividal** [comment]
No, that should be it. The CLLI number is indeed all we need.

![image](https://github.com/user-attachments/assets/bbd5bac3-d090-4aa6-8905-3d77e5fc45fa)

**Comment by vassbo** [comment]
Here you can try it out: https://github.com/ChurchApps/FreeShow/releases/tag/v1.2.9
To export, go to settings>"Other">"Export usage log".

Let me know if you have updates on the uploader script.

**Comment by yurividal** [comment]
@vassbo 
I managed to reverse engineer CCLIs API and found a way to programatically report a song. 
I put together a quick code in python as proof of concept. I tested with a couple of songs, and they all show in my CCLI history as reported.
The API is a bit more tricky on the AUTH part, since they include captchas etc...

Perhaps, we can have a webview login page in freeshow, which can then grab the cookies and use them for submittting the reports.

I don't know if can post it here, or if you would rather have this sent by email. 
just let me know

**Comment by vassbo** [comment]
Nice! For me it's the same.

I think @gwonamfromkoradai was working on a login system?: https://github.com/ChurchApps/FreeShow/issues/572#issuecomment-2399873916

Let's see if we can merge these in a way, both song importing & usage reporting!

**Comment by gwonamfromkoradai** [comment]
Yes I am! Working out a few kinks though. The login is fine but I’m just trying to make the search/save function worm

**Comment by gwonamfromkoradai** [comment]
I like the idea of a web view page. I had an annoying time dealing with the two cookie pop ups (who does two??) alone with a popup that says unsupported when I run selenium headless mode.

**Comment by yurividal** [comment]
Great!
@gwonamfromkoradai 
Are you able to obtain these cookies from your login process?

```
ARRAffinity
ARRAffinitySameSite
.AspNetCore.Session
.AspNetCore.Antiforgery
CCLI_AUTH
CCLI_JWT_AUTH
```

and also
`RequestVerificationToken`


If we have these values, my script is already able to report songs.
Currently, i open a browser session, login, and get these values from developer options.

**Comment by yurividal** [comment]
@vassbo @gwonamfromkoradai 

[Here](https://github.com/yurividal/ccli_auto) is my code. Its not pretty, but it works as a proof of concept.

It automatically spawns a chrome page, and tries to log in. Once logged in, it will need to submit a single CCLI song, in order to capture the RequestverificationToken and the Cookies. 
Once we have that info, it can then use the API to bulk report all the rest of the list.

Works as a proof of concept, and might give us some ideas of how it can be implemented in freeshow.

**Comment by gwonamfromkoradai** [comment]
Awesome. That’s similar to what I was doing but your APk integration makes it a lot smoother. 

When I’m back at my computer, I’ll try integrating the search/add function into this, that way we can have one going (unless that’s not a good idea?). I’m off for a couple days but should be able to write something up Tuesday or Wednesday. 

I think that this code has some missing pieces to help simplify what my code was doing. I might have the first version just save them as txt files for now, then once it is working, make it work as a plugin. 

Also if anyone else solves the search stuff before me I won’t be offended lol. Otherwise, I’ll work on it in a few days. 

As a side note, if anyone wanted a workaround for now on Mac, I built an Automator which imports to freeshow using the free version of ProPresenter lol. Not a permanent solution and the definition of over-engineered but it works. But tbh at that point you might as well use the web interface / copy and paste.

**Comment by yurividal** [comment]
I found a way to ge tthe verification token without having to manually submit one song. I've updated the script

**Comment by vassbo** [comment]
Looks good, thank you for all the help! 🎉

**Comment by yurividal** [comment]
@vassbo i've just submited 101 songs on a single report, and it worked perfectly. I was concerned that the API might have some sort of restriction on how many you can do at a time. Now we know that at least 100 songs can be reported in a single post.

I've also coded a "delete" script, that will delete any reported songs in the last 3 monts. (Needed a way to retract all the 100 test songs i reported. hahaha)

**Comment by Marsipulami** [comment]
Love it that you all are still working on SongSelect features. How is this proces going and is it correct, that if i am reading this correct, that https://github.com/ChurchApps/FreeShow/issues/776 maybe integrated after all ?

**Comment by vassbo** [comment]
@gwonamfromkoradai Did you get a chance to work on the song search & importing?

**Comment by gwonamfromkoradai** [comment]
Hey, I was working on it but have been out of commission for a bit. I'm going to look at the code I have and see what might work.

**Comment by vassbo** [comment]
Okay no worries, take your time!

**Comment by gwonamfromkoradai** [comment]
FWIW if anyone gets to this before me, there will be no offence taken at all lol.

I think I will upload my garbage code that I have and then keep working on it, but that way if other people can integrate it better they can.

**Comment by gwonamfromkoradai** [comment]
@yurividal I've been working on my old code but I think your code is maybe a better starting place. I was using selenium and navigating the site, but I wonder if utilizing your program's ability to grab the cookie and request verification token may be a better avenue to go down in order to make it cleaner. 

Time for some dumb questions from me. Does the verification token = access token? Does your code extract the subscription key? 

If that wouldn't work here, another option is for me to adapt my old garbage code into a chrome extension or something, basically use the CCLI website and have ability to import directly from that. Not as clean as what you've made for reporting, however.

**Comment by yurividal** [comment]
I dont think i grab the subscription key. What is that used for? I don't think i saw any refference to subscription keys when i was inspecting the traffic

**Comment by yurividal** [comment]
Ok, from the [documentation here](https://documenter.getpostman.com/view/604633/TzseGkmA#auth-info-bb8af6b4-12d6-4d26-8500-86c8d4517c61) the subscription key is a key provided by CCLI when you get your CCLI Partner account. I don't think we have that, and the whole point of this discussion is to be able to use their public API without this partner account, which aparently costs over $1K/year

If we want to implement songselect search, we will have to utilize the auth and api calls implemented on https://songselect.ccli.com/

**Comment by klecu** [comment]
Could you do like [OpenSongApp](https://github.com/thebigg73/OpenSongTablet) and add a webview that intercepts .chopro and .txt downloads?

**Comment by yurividal** [comment]
The more  i look into the code on their website, the more complicated and protected it looks.
I think the easiest wya to implement the songselect import would be to:
- Open a browser window when user clicks import SongSelect
- Let the user login, and search for the song
- User clicks the download button
- Freeshow waits and expects a txt file
- Once txt file is found, freeshow parses it

**Comment by gwonamfromkoradai** [comment]
Yeah that’s exactly what I was thinking. That’s why I’m wondering if a chrome extension might work. 

I’m looking through the openlp code as well to see how they handle it. 

I didn’t realize there was a price to access the API, I thought they were just not given it out except to the elect few. That makes even less sense. I get why pro presenter doesn’t want apps like this to have import features, but I honestly don’t get why CCLI would want to prevent it. Seems shady.

**Comment by yurividal** [comment]
Everything is shady about CCLI... Unfortunately

**Comment by gwonamfromkoradai** [comment]
I may have something, got some inspiration from open lp’s design. I did a rough program, I want to try to combine it with the importer

**Comment by gwonamfromkoradai** [comment]
Ok, here is what I've got. Not much but maybe a start?

https://github.com/gwonamfromkoradai/SongSelectSave/tree/main

It's a simple python script that will open a browser window to CCLI, allow you to log in, then when you search, you can select copy and you can save it as a text file. 

Technically speaking., you can use this to import from CCLI to FreeShow by then using the text import feature and pointing to your "song" directory. However, I see this more as a first step to adapt a converter - I don't think that it would take much to adapt this directly within the existing import function. I was working on adapting this to a .ts that could be integrated but I didn't get too far yet. 

I also would guess that this could be easily integrated with what @yurividal developed for the reporting software so that users will only need to sign in once. In fact, it probably would make this code more easily integrated into the program. 

Let me know thoughts. I am happy to keep chipping away but wanted to share what I got so far. Sorry it's taken so long for not much lol, I've tried to integrate other features but ended up stripping it down to the bare features to (I think) provide a framework to integrate Song Select search into FreeShow.

Lastly, I tried but couldn't for the life of me accomplish this without a web interface pop-up. Perhaps smarter people than me might be able, but I did notice that OpenLP has a similar set-up to this (They utilize the download function, but in my humble opinion the clipboard grabber is cleaner because of the way Song Select is laid out with the copy function). The main thing I couldn't get working was the search function, but again, maybe with @yurividal's success utilizing the reporting search, a similar thing might be adapted.

**Comment by vassbo** [comment]
Nice! Looks like the repo is private.

The best would be to make this into a node module/npm package, to be able to easily integrate it into the program.

**Comment by gwonamfromkoradai** [comment]
Ok, here is a node module version:

https://github.com/gwonamfromkoradai/SSSave/releases/tag/SSSave

I think my next step will be adding file conversion to make the files save as .show 

After that, I would like to try to clean up the keys. Then at the end, I will tinker with the UI.

**Comment by gwonamfromkoradai** [comment]
Hello all!

Sorry for the very long delay. I think I finally have a beta program that could effectively work as a baseline to integrate SongSelect search into Freeshow!

[Here is the code. ](https://github.com/gwonamfromkoradai/SongSelectSave/tree/main) I am going to work on making a node module version which would (hopefully) make it more easily integrated into Freeshow. 

Basically, the first time you use it, it will ask for your CCLI login. Then it will launch a GUI search function, where you can search, select a song, and save it as a txt file. 

One thing that I am happy about is that it is able to function entirely within the GUI on the user-level, using a web backend. The downside of this approach, of course, is that it will break if CCLI changes their site element labels. But it wouldn't be a huge deal to update them in that case. 

Thanks to the work of Yurividal, I was also able to integrate what he developed into this allowing the user to be able to remain logged in using a saved cookie/token. You only will need to login once. 

The code is clunky and I know it will need tweaking. For example, I don't know if changing one's password changes their access token? Also, I am wondering about changing it so that the username/password isn't saved in a file, since they will have the cookie/token saved anyways. It's also not the speediest thing ever made, I would imagine there are ways that it could be improved. Nonetheless, I have to admit I am happy to see a proof of concept of a fully GUI song search/save with songselect. I am also guessing that there will be ways that integration will make it work smoother. 

This lacks the converter function. I was working on it, but I gave up because I was thinking that something could probably be made using the existing text file import function? IE on the front end it is just importing song select, but on the back end it is running this type of program, saving as texts, then running the text to .show converter? Or maybe there was an easier way.

**Comment by jschwalbe** [comment]
@gwonamfromkoradai nice! I'm *just* now looking into transitioning my (small) church from Pr*presenter to FreeShow and SongSelect was one of the questions that came up. 

Is there a way to test this within FreeShow or just try it out from the command line? What's the next step?

**Comment by fpbcyouthfellowship-wq** [comment]
Please, I am currently having a grey out web search for lyrics in New Show tab
What can I do?

**Comment by vassbo** [comment]
@fpbcyouthfellowship-wq Type a search name.

**Comment by gwonamfromkoradai** [comment]
Sorry for the edge of the earth I fell off. Let me know if there's anything I can do, I kind of paused any work on this.

**Comment by vassbo** [comment]
No worries @gwonamfromkoradai. If you got time for the node module that would be nice, but if not thats okay.

**Comment by gwonamfromkoradai** [comment]
Sounds good. My old code has broken as it relied on the website to not make weird changes, so I am remaking it based on a backend around how the actual website interacts with the API. Seems to be working pretty well so far. 

If I can make this work, it should be a lot smoother to integrate so it doesn't have to be constantly breaking depending on the website.

**Comment by yurividal** [comment]
So, it seems that CCLI has enabled cloudflare captcha on the login process, which broke my previous automation for logging in and obtaining the cookies.

I have adapted my script for it. It now, launches a google chrome screen, and the user logs in manually. Once the user logs in the first time, we capture the session cookie and token, which can then be used for future reports, until the token expires. 

The script also now parses the freeshow report json. It searches for any CCLI song number, queries CCLI for the song ID and title. Then, it saves all the info into a local cache (so we don't see to query known songs again). Once it got the details of all the songs in the report, it finally reports the usage of the songs in the list.

Its written in python, but it works pretty well.
I just reported over 200 songs, from the past 8 months, in just a few seconds. 

Code here: https://github.com/yurividal/ccli_auto

**Comment by nilber79** [comment]
Is the any current effort to integrate the auto reporting functionality into FreeShow? Has anyone reached out to CCLI to discuss a partnership to allow direct API integration like some of the other platforms have?

The Web Search function is there but doesn't seem to pull any results from CCLI SongSelect, is this still an ongoing effort as well?

**Comment by yurividal** [comment]
I think the issue here is more ethical than technical.
As mentioned, I have a script that totally automates the reporting for me, based on the export from freeshow.
But, implementing it and integrating it into freeshow could cause some issues and backslash from CCLI. It might violate their license, and they might get angry at freeshow.
So, for now, I think it's best to keep it as a separate project. Feel free to test, and submit PRs to my repo if you want.

**Comment by vassbo** [comment]
> Has anyone reached out to CCLI to discuss a partnership to allow direct API integration like some of the other platforms have?

Yes, we have, and they are not willing to collaborate. so it's unfortunate: https://github.com/ChurchApps/FreeShow/issues/572

**Comment by nilber79** [comment]
As near as I can tell CCLI has two separate pieces here, both identified in the title of this issue.

1.) SongSelect - CCLI deprecated the API for doing song searches. Issue #572 addresses this. Got it, not going to happen. Given that anyone that wants to take advantage of SongSelect has to have a paid subscription through CCLI, and not Renewed Vision, they reasonably should have access in other software, like FreeShow, to the information. I'd love to see if there is anything in the CCLI agreement that says SongSelect can't be used with other software.

2.) CCLI Usage Reporting - Addressed by @yurividal's project, but not addressed in issue #572. I don't understand why they would make it as difficult as possible to report usage. There is apparently no reporting API, which I really can't understand, as a software developer myself. It takes minimal effort to incorporate an API into your backend if you're already processing data through a website, and storing it in a database.

@yurividal I can't understand the backlash from using the same mechanism that Renewed Vision seems to be using, for either function. Unless the CCLI Terms of Use specifically forbid it and specifically allow Renewed Vision users an exception, there isn't much CCLI can do if you are a paying user and presenting a valid authentication token to access or report the information. If there were an expectation from FreeShow, or another developer, that CCLI specifically modify  CCLI code to accommodate another piece of software, I could see CCLI not doing that. As long as the impetus remains with the second party developer to adapt to the CCLI platform, and they are allowing at least one second party (Renewed Vision) then why not allow others.

I'd be willing to champion the cause with CCLI, but if the horse is already dead, then there's no reason. From looking at the response in issue #572, it seems like one response from CCLI stating the SongSelect API was deprecated was taken, by the FreeShow developers, as a final answer and any integration effort was abandoned. I understand there would be an ongoing maintenance piece, for FreeShow, to re-evaluate changes to the CCLI front-end to maintain the integration. I would also inquire from CCLI, why they can't be more transparent with all developers and at least provide a changelog if they do make changes that could break second party integrations. I understand they may treat some of it as proprietary information, but at least give other developers that aren't Renewed Vision a somewhat equal footing.

My thoughts on implementation are:
1.) Identify CCLI endpoints - mostly done by @yurividal 
2.) Utilize the Electron native `session/webContents` API to host a real embedded browser window for the CCLI login page, then read the resulting cookies directly from Electron's cookie store once login succeeds.
3.) Persist and refresh the session. Store the session cookie + anti-forgery token encrypted at rest using Electron's safeStorage API
4.) Replay requests for search and reporting. Once authenticated, hit the same internal endpoints the browser hits for SongSelect search and for submitting a usage report, formatted the same way.

What is the main reason for FreeShow not to move forward with CCLI integration?
1.) Maintenance
2.) Misunderstanding of the CCLI response to issue #572 
3.) Lack of desire to integrate

**Comment by yurividal** [comment]
I think the main reason why they don't make the API available is that they want software to partner with them to do it. I think I remember reading somewhere that they charge about 1200 USD/year for access to the API. That's how Pro presenter and some other software manage to have automatic CCLI reporting - they are paid software, and can incorporate the cost of the API access into their subscription price.
It's all about the $$$

**Comment by yurividal** [comment]
Continuing...
My concern is that, if we integrate this into freeshow, they might realize that many people are using it, and actively add more mechanisms to stop us.
They did this a while ago. When we first started automating the reports, they implemented a very restrictive CAPTCHA.

**Comment by gwonamfromkoradai** [comment]
Because of the reporting script created before I ever entered this world, my interest was primarily in song import. I personally find it frankly putrid that there would be company specific restrictions imposed by CCLI / songselect, and my codes were an attempt to try to automate this. Theoretically, there is little to stop this from being directly integrated into Freeshow, but I think there is also a pragmatic aspect at play here.

For me, I'm a very crappy coder, so I basically built a stand alone program and left it. Others can speak to the reporting aspect better than me. However, even after coding a standalone program (which I believe has been forked and improved upon), my own workflow for importing songs from songselect has been

1. Search for song on song select web browser
2. Use copy function in song select
3. Go to Freeshow, import from clipboard. 

This is what in my own workplace I've trained volunteers to do, and I wonder if perhaps we could integrate this directly into Freeshow's interface to make it simpler. 

Anyway, I admit I am going off memory and may need to revise my ramblings after reviewing what has been done. 

Edit @maximilianbiebl did a fork of my initial code and improved on it. https://github.com/maximilianbiebl/ccli_downloader/

**Comment by gwonamfromkoradai** [comment]
Yes, it seems that even if it doesn't actually break terms of service, they are incentivized to update the API in a way that breaks the ability for scripts like this to work.

> Continuing... My concern is that, if we integrate this into freeshow, they might realize that many people are using it, and actively add more mechanisms to stop us. They did this a while ago. When we first started automating the reports, they implemented a very restrictive CAPTCHA.

**Comment by gwonamfromkoradai** [comment]
> The more i look into the code on their website, the more complicated and protected it looks. I think the easiest wya to implement the songselect import would be to:
> 
> * Open a browser window when user clicks import SongSelect
> * Let the user login, and search for the song
> * User clicks the download button
> * Freeshow waits and expects a txt file
> * Once txt file is found, freeshow parses it

Resurrecting this old comment to add - would it simplify or complicate the flow if rather than txt this is directly integrated with existing clipboard import function

**Comment by MarkJepson** [comment]
I too was concerned about the integration, However, Freeshow actually works well with copy and paste

- Log into SongSelect
- Go to the lyrics page for the song
- Press the copy **key** - Do not highlight and copy
- Paste into FreeShow's Quick Lyrics

This gives FreeShow everything it needs to display copyright and licence information, as well as the CCLI song number for future reporting.

**Comment by gwonamfromkoradai** [comment]
Yes, that’s how I do it as well. It’s pretty easy to do and there is no concern at all about CCLI terms. 

I am wondering about a button in the program or official support which provides instructions? I’ve manually taught people this and it’s very simple, a big reason I gave up on updating my search program

> I too was concerned about the integration, However, Freeshow actually works well with copy and paste
> 
> * Log into SongSelect
> * Go to the lyrics page for the song
> * Press the copy **key** - Do not highlight and copy
> * Paste into FreeShow's Quick Lyrics
> 
> This gives FreeShow everything it needs to display copyright and licence information, as well as the CCLI song number for future reporting.

**Comment by vassbo** [comment]
@gwonamfromkoradai Updated the info: https://freeshow.app/docs/faq#ccli-integration
